### PR TITLE
Bug1277038 - Differ between `oc env` updates(removes/adds) the resource(s) envvars and when it doesnt 

### PR DESCRIPTION
**Pull request by jhadvig**
When adding/removing envvars with `oc env` cmd, differ between states when the command actually updates the resource(s) and when it doesnt.
Eg: When updating resource by adding the envvar with a value that is already present in the resource, or removing a envvar that is not present, the printed output will look accordingly:
`replicationcontroller "database-1" nothing to update`
instead of 
`replicationcontroller "database-1" updated`

+ added a TC

Wasnt sure if we also wanna print due to which envar the resource wasnt updated (eg1. removing envvar thats not present in the resource || updating envar with the same value), since we would have to revrite some parts of the `env.go` and will end up with dup. messages in case user would want to update all rcs/dc with wrong envars(eg1).

@fabianofranz thoughts ?

### CODE REVIEW

**Comment by fabianofranz** [comment]
[test]

**Comment by liggitt** [comment]
Will conflict/overlap with https://github.com/openshift/origin/pull/5819
How do we treat deletions of resources which don't exist?

**Comment by jhadvig** [comment]
@liggitt I've already spoken with @kargakis about the overlap and this PR doesn't solve this issue.
So we could either:
- merge this and rebase https://github.com/openshift/origin/pull/5819
- wait until https://github.com/openshift/origin/pull/5819 is merged and rebase this PR

Don't have any strong feelings about both solutions. Thoughts ?

**Comment by kargakis** [comment]
#5819 supersedes this. The success message there needs fixing though, @jhadvig may want to rebase your changes regarding it once it gets merged.

**Comment by jhadvig** [comment]
@kargakis yes, once the our PR is merged, will rebase. Thx

**Comment by jhadvig** [comment]
@fabianofranz I've updated the PR after #5819 got merged
Now when deleting envs that aren't present in the pod's container the `oc` will return:
```
$ oc env dc/frontend TEST1- TEST2-
environment variable "TEST1" not found in ruby-helloworld container
environment variable "TEST2" not found in ruby-helloworld container
deploymentconfig "frontend" nothing to update
``` 

PTAL

**Comment by liggitt** [comment]
What do the label/annotate commands do in this case?

**Comment by jhadvig** [comment]
@liggitt I've already proposed this kind of change for the `label` cmd to upstream and it was merged aswell https://github.com/kubernetes/kubernetes/pull/17946

Regarding the `annotate` cmd, its not returning this kind of info(for now)

File: pkg/cmd/cli/cmd/env.go [hunk 1]
```diff
@@ -260,7 +265,19 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
 				return nil
 			}
 			for _, c := range containers {
+				oldEnv := c.Env
+				for _, env := range remove {
+					_, ok := findEnv(c.Env, env)
+					if !ok {
+						infosMsgs[i] = append(infosMsgs[i], fmt.Sprintf("environment variable %q not found in %s container\n", env, c.Name))
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: So if something has multiple containers, one of which has an envvar, and you remove the envvar, you'll always get a message saying the other container didn't have that var? That doesn't seem necessary.

File: pkg/cmd/cli/cmd/env.go [hunk 2]
```diff
@@ -260,7 +265,19 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
...
+					}
+				}
 				c.Env = updateEnv(c.Env, env, remove)
+				if reflect.DeepEqual(oldEnv, c.Env) {
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: This appears to set the updated flag based on the last container, so changes to earlier containers would be lost if the last container didn't change, right?

Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: I would expect the API call to do the update to be skipped if no changes were needed

File: pkg/cmd/cli/cmd/env.go (hunk 1 above)
Comment (jhadvig) [inline pkg/cmd/cli/cmd/env.go]: if user is updating the envars in a pod, by default it will update all the containers. For that reason user should specify `--containers` flag I guess.

Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: I wouldn't print warnings that an envvar they wanted to remove didn't exist in all containers. If it existed in at least one, that's probably fine

Comment (jhadvig) [inline pkg/cmd/cli/cmd/env.go]: Hmm, how about printing them with glog at some higher verbosity ?

Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: nah, if I ask to remove envvar B, and it exists in one container but not another, I don't consider that failure or warning-worthy

File: pkg/cmd/cli/cmd/env.go (hunk 2 above)
Comment (jhadvig) [inline pkg/cmd/cli/cmd/env.go]: 1, right
2, you mean the PATCH call?

Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: Yeah, no API call if no changes were made

**Comment by jhadvig** [comment]
@liggitt updated
ptal

File: pkg/cmd/cli/cmd/env.go [hunk 3]
```diff
@@ -324,24 +333,28 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
...
-			return err
-		}
-		obj, err := resource.NewHelper(info.Client, info.Mapping).Patch(info.Namespace, info.Name, kapi.StrategicMergePatchType, patchBytes)
-		if err != nil {
-			handlePodUpdateError(cmd.Out(), err, "environment variables")
-			failed = true
-			continue
+		outputMsg := "updated"
+		if !updated[i] {
+			outputMsg = "nothing to update"
```
Comment (kargakis) [inline pkg/cmd/cli/cmd/env.go]: Instead of doing this and nesting all the code, do this:
```go
if !updated[i] {
   cmdutil.PrintSuccess(mapper, shortOutput, out, info.Mapping.Resource, info.Name, "didn't need to update")
   continue
}
// rest of the code...
```

Comment (kargakis) [inline pkg/cmd/cli/cmd/env.go]: Also it would be nice if this (no need for update) was handled by `CreateTwoWayMergePatch` by returning an empty byte slice and not by us but that's not the case: https://github.com/kubernetes/kubernetes/issues/17022

Comment (kargakis) [inline pkg/cmd/cli/cmd/env.go]: Actually @liggitt if we pin the size of an empty patch to a constant and check for it, wouldn't it be better?

Comment (jhadvig) [inline pkg/cmd/cli/cmd/env.go]: The size of empty patch `2`

Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: It's `{}`, right? Not sure how I feel about checking that.  I think I would rather have `updateEnv` additionally return a boolean indicating whether any updates were made

**Comment by jhadvig** [comment]
@liggitt @kargakis I've updated the PR. `updateEnv` will additionally return bool whether container envvars had been updated
ptal

File: pkg/cmd/cli/cmd/env.go [hunk 4]
```diff
@@ -371,7 +384,11 @@ func updateEnv(existing []kapi.EnvVar, env []kapi.EnvVar, remove []string) []kap
 		covered.Insert(e.Name)
 		out = append(out, e)
 	}
-	return out
+	envUpdated := true
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: instead of using reflection, I'd rather initialize envUpdated to false at the top, and set it to true in the loop where we know we're making changes

Comment (jhadvig) [inline pkg/cmd/cli/cmd/env.go]: Then the line which would be setting `envUpdate` to false would need to be basically in all three `if` statements and also at the end of the `for` loops, due to the logic the `out` variable is build. Would rather prevent prevent that.

Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: I still would rather set it explicitly... I think some of the complexity would be reduced by handling each case (removed, modified, added) in its own loop

Comment (jhadvig) [inline pkg/cmd/cli/cmd/env.go]: ok, will rewrite the logic and set it explicitly :)

**Comment by jhadvig** [comment]
@liggitt cleaned the logic a bit as suggested.
ptal

File: pkg/cmd/cli/cmd/env.go [hunk 5]
```diff
@@ -252,7 +254,10 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
 	}
 
 	skipped := 0
-	for _, info := range infos {
+	var envUpdated bool
+	var updated = make([]bool, len(infos))
+	for i, info := range infos {
+		updated[i] = false
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: defaults to `false` alreayd

File: pkg/cmd/cli/cmd/env.go [hunk 6]
```diff
@@ -260,7 +265,10 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
 				return nil
 			}
 			for _, c := range containers {
-				c.Env = updateEnv(c.Env, env, remove)
+				c.Env, envUpdated = updateEnv(c.Env, env, remove)
+				if envUpdated {
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: `updated[i] = updates[i] || envUpdated`

File: pkg/cmd/cli/cmd/env.go [hunk 7]
```diff
@@ -324,6 +332,11 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
 
 	failed := false
 	for i, info := range infos {
+		shortOutput := cmdutil.GetFlagString(cmd, "output") == "name"
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: move `shortOutput := cmdutil.GetFlagString(cmd, "output") == "name"` outside the loop

File: pkg/cmd/cli/cmd/env.go [hunk 8]
```diff
@@ -349,36 +361,38 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
...
+			envUpdated = true
 			continue
 		}
-		newer, ok := findEnv(env, e.Name)
+		newer, ok := findEnv(env, e)
 		if ok {
-			covered.Insert(e.Name)
-			out = append(out, newer)
-			continue
+			envUpdated = true
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: only if the value has changed, right?

Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: I guess this is probably a tight enough check

File: pkg/cmd/cli/cmd/env.go [hunk 9]
```diff
@@ -349,36 +361,38 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
 	return nil
 }
 
-func updateEnv(existing []kapi.EnvVar, env []kapi.EnvVar, remove []string) []kapi.EnvVar {
+func updateEnv(existing []kapi.EnvVar, env []kapi.EnvVar, remove []string) ([]kapi.EnvVar, bool) {
 	out := []kapi.EnvVar{}
-	covered := sets.NewString(remove...)
+	envUpdated := false
+	toDelete := sets.NewString(remove...)
+
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: need to track existing names in a set

File: pkg/cmd/cli/cmd/env.go [hunk 10]
```diff
@@ -349,36 +361,38 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
...
-			continue
+			envUpdated = true
 		}
-		out = append(out, e)
+		out = append(out, newer)
 	}
+
 	for _, e := range env {
-		if covered.Has(e.Name) {
+		if toDelete.Has(e.Name) {
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: should be skipping if the env already exists, not if it was in the set to delete... it's an error to delete and set/update the same envvar in the same command (prevented by validation, I'm pretty sure)

File: pkg/cmd/cli/cmd/env.go [hunk 11]
```diff
@@ -349,36 +361,38 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
...
+		envUpdated = true
 	}
-	return out
+	return out, envUpdated
 }
 
-func findEnv(env []kapi.EnvVar, name string) (kapi.EnvVar, bool) {
-	for _, e := range env {
-		if e.Name == name {
+func findEnv(envs []kapi.EnvVar, env kapi.EnvVar) (kapi.EnvVar, bool) {
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: why the change from looking for a name?

File: pkg/cmd/cli/cmd/env.go [hunk 12]
```diff
@@ -349,36 +361,38 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
...
+	for _, e := range envs {
+		if e.Name == env.Name {
 			return e, true
 		}
 	}
-	return kapi.EnvVar{}, false
+	return env, false
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: this is confusing... I prefer the original impl where a `false` retval meant you should ignore the returned env

File: pkg/cmd/cli/cmd/env.go (hunk 6 above)
Comment (jhadvig) [inline pkg/cmd/cli/cmd/env.go]: all the items in `updated` array are set to `false` by default, so no point to check for it, just put the `envUpdated` value to it.

File: pkg/cmd/cli/cmd/env.go (hunk 12 above)
Comment (jhadvig) [inline pkg/cmd/cli/cmd/env.go]: the idea is to return the new `EnvVar` in case of overriding the envvar or return the old one

File: pkg/cmd/cli/cmd/env.go (hunk 11 above)
Comment (jhadvig) [inline pkg/cmd/cli/cmd/env.go]: comment below

File: pkg/cmd/cli/cmd/env.go (hunk 6 above)
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: nope, then the last one in wins... need to `||`

File: pkg/cmd/cli/cmd/env.go (hunk 12 above)
Comment (liggitt) [inline pkg/cmd/cli/cmd/env.go]: the old way was clearer... putting a value in and not easily knowing which you get back is confusing

File: pkg/cmd/cli/cmd/env.go (hunk 6 above)
Comment (jhadvig) [inline pkg/cmd/cli/cmd/env.go]: but in this case it doesn't matter which wins. Once the `updated[i]` is set to `true` it can't be changed in any iteration.

**Comment by danmcp** [comment]
@jhadvig Bump

**Comment by danmcp** [comment]
Closing due to age and inactivity.  Feel free to reopen if you are going to continue working on it.
